### ISSUE TITLE
`group_by.agg` slice in v0.19.1 returns wrong output (dtype and values)

### ISSUE THREAD

**Issue by topher-lo** [pola-rs/polars#10897]
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl

X = pl.scan_parquet("https://github.com/neocortexdb/functime/raw/main/data/commodities.parquet")
test_size = 6
entity_col = X.columns[0]
train_split = (
    X.groupby(entity_col)
    .agg(pl.all().slice(0, pl.count() - test_size))
    .explode(pl.all().exclude(entity_col))
)
test_split = (
    X.groupby(entity_col)
    .agg(pl.all().slice(-1 * test_size, test_size))
    .explode(pl.all().exclude(entity_col))
)
print(train_split.collect())  # Wrong ("time" and "price" get coerced into uint32)
print(test_split.collect())  # Expected output
```

### Log output

Got the following in v0.19.1

```shell
keys/aggregates are not partitionable: running default HASH AGGREGATION
shape: (47_157, 3)
┌────────────────┬───────┬───────┐
│ commodity_type ┆ time  ┆ price │
│ ---            ┆ ---   ┆ ---   │
│ str            ┆ u32   ┆ u32   │
╞════════════════╪═══════╪═══════╡
│ Logs, Cameroon ┆ 47577 ┆ 47577 │
│ Logs, Cameroon ┆ 47577 ┆ 47577 │
│ Logs, Cameroon ┆ 47577 ┆ 47577 │
│ Logs, Cameroon ┆ 47577 ┆ 47577 │
│ …              ┆ …     ┆ …     │
│ Silver         ┆ 47577 ┆ 47577 │
│ Silver         ┆ 47577 ┆ 47577 │
│ Silver         ┆ 47577 ┆ 47577 │
│ Silver         ┆ 47577 ┆ 47577 │
└────────────────┴───────┴───────┘
```

### Issue description

Groupby slice operation on a LazyFrame returns unexpected result (coerces into `u32` and fills constant value into every cell). This only occurs if `length=pl.count() - some_int` in `pl.col("col_name").slice`

Discovered this bug when I tried to bump polars from `0.18.11` to `0.19.1` in `functime`. In particular, the code snippet above comes from functime's [`train_test_split`](https://github.com/neocortexdb/functime/blob/main/functime/cross_validation.py#L7) function

### Expected behavior

Expected (working in `polars` v0.18.11)
```
keys/aggregates are not partitionable: running default HASH AGGREGATION
shape: (47_157, 3)
┌─────────────────────┬─────────────────────┬────────┐
│ commodity_type      ┆ time                ┆ price  │
│ ---                 ┆ ---                 ┆ ---    │
│ str                 ┆ datetime[ns]        ┆ f64    │
╞═════════════════════╪═════════════════════╪════════╡
│ Potassium chloride  ┆ 1960-01-01 00:00:00 ┆ 28.5   │
│ Potassium chloride  ┆ 1960-02-01 00:00:00 ┆ 28.5   │
│ Potassium chloride  ┆ 1960-03-01 00:00:00 ┆ 28.5   │
│ Potassium chloride  ┆ 1960-04-01 00:00:00 ┆ 28.5   │
│ …                   ┆ …                   ┆ …      │
│ Wheat, US HRW       ┆ 2022-06-01 00:00:00 ┆ 459.59 │
│ Wheat, US HRW       ┆ 2022-07-01 00:00:00 ┆ 382.5  │
│ Wheat, US HRW       ┆ 2022-08-01 00:00:00 ┆ 382.86 │
│ Wheat, US HRW       ┆ 2022-09-01 00:00:00 ┆ 419.14 │
└─────────────────────┴─────────────────────┴────────┘
```

### Installed versions

<details>

```
--------Version info---------
Polars:              0.19.1
Index type:          UInt32
Platform:            Linux-5.15.90.1-microsoft-standard-WSL2-x86_64-with-glibc2.10
Python:              3.8.17 | packaged by conda-forge | (default, Jun 16 2023, 07:06:00)
[GCC 11.4.0]

----Optional dependencies----
adbc_driver_sqlite:  <not installed>
cloudpickle:         2.0.0
connectorx:          <not installed>
deltalake:           <not installed>
fsspec:              2023.6.0
matplotlib:          3.7.2
numpy:               1.24.4
pandas:              2.0.3
pyarrow:             12.0.1
pydantic:            2.0.3
sqlalchemy:          2.0.19
xlsx2csv:            <not installed>
xlsxwriter:          <not installed>
```

</details>

**Comment by ritchie46** [comment]
Curious what change introduced this. Anybody time for a bisect on this one?

**Comment by cmdlineluser** [comment]
@ritchie46 It stopped working in `0.18.14`

It works with `comm_subexpr_elim=False` so I guess it's related to https://github.com/pola-rs/polars/commit/1b0767c0d85a89323758a5eebfb4c7fd1b575a16

```python
>>> train_split.collect(comm_subexpr_elim=False)
shape: (47_157, 3)
┌────────────────┬─────────────────────┬───────┐
│ commodity_type ┆ time                ┆ price │
│ ---            ┆ ---                 ┆ ---   │
│ str            ┆ datetime[ns]        ┆ f64   │
╞════════════════╪═════════════════════╪═══════╡
│ Logs, Cameroon ┆ 1970-01-01 00:00:00 ┆ 42.99 │
│ Logs, Cameroon ┆ 1970-02-01 00:00:00 ┆ 42.99 │
│ Logs, Cameroon ┆ 1970-03-01 00:00:00 ┆ 42.99 │
│ Logs, Cameroon ┆ 1970-04-01 00:00:00 ┆ 42.99 │
│ …              ┆ …                   ┆ …     │
│ Sugar, world   ┆ 2022-06-01 00:00:00 ┆ 0.42  │
│ Sugar, world   ┆ 2022-07-01 00:00:00 ┆ 0.4   │
│ Sugar, world   ┆ 2022-08-01 00:00:00 ┆ 0.39  │
│ Sugar, world   ┆ 2022-09-01 00:00:00 ┆ 0.39  │
└────────────────┴─────────────────────┴───────┘
```

**Comment by ritchie46** [comment]
This was a false positive in the new CSE optimizer. Fix coming up.

**Comment by ritchie46** [comment]
fixed by #10917
